lib-multisrc/mangahub/src/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt

```diff
@@ -182,39 +184,72 @@ abstract class MangaHub(
-            } catch (_: IOException) {
-                throw IOException("An error occurred while obtaining a new API key") // Show error
+    private val lock = ReentrantLock()
+    private var refreshed = 0L
+
+    private fun refreshApiKey(manga: SManga? = null, chapter: SChapter? = null) {
+        if (refreshed + 10000 < System.currentTimeMillis() && lock.tryLock()) {
+            val url = when {
+                chapter != null -> "$baseUrl/chapter${chapter.url}"
+                manga != null -> "$baseUrl${manga.url}"
+                else -> "$baseUrl/chapter/martial-peak/chapter-${Random.nextInt(2000, 3000)}"
+            }.toHttpUrl()
+
+            val oldKey = client.cookieJar
+                .loadForRequest(baseUrl.toHttpUrl())
+                .firstOrNull { it.name == "mhub_access" && it.value.isNotEmpty() }?.value
+
+            for (i in 1..2) {
+                // Clear key cookie
+                val cookie = Cookie.parse(url, "mhub_access=; Max-Age=0; Path=/")!!
+                client.cookieJar.saveFromResponse(url, listOf(cookie))
+
+                try {
+                    // We try requesting again with param if the first one fails
+                    val query = if (i == 2) "?reloadKey=1" else ""
+                    val response = client.newCall(
+                        GET(
+                            "$url$query",
+                            headers.newBuilder()
+                                .set("Referer", "$baseUrl/manga/${url.pathSegments[1]}")
+                                .build(),
+                        ),
+                    ).execute()
+                    val returnedKey =
+                        response.headers["set-cookie"]?.let { apiRegex.find(it)?.groupValues?.get(1) }
+                    response.close() // Avoid potential resource leaks
+
+                    if (returnedKey != oldKey) break // Break out of loop since we got an allegedly valid API key
+                } catch (_: Throwable) {
+                    lock.unlock()
+                    throw Exception("An error occurred while obtaining a new API key") // Show error
+                }
             }
+            refreshed = System.currentTimeMillis()
+            lock.unlock()
+        } else {
+            lock.lock() // wait here until lock is released
+            lock.unlock()
         }
     }
 
+    private inline fun <reified T> fetchAndRetry(
+        manga: SManga? = null,
+        chapter: SChapter? = null,
+        crossinline fetch: () -> Observable<T>,
+    ): Observable<T> {
+        return fetch()
+            .onErrorResumeNext { error ->
+                if (errorRegex.containsMatchIn(error.message ?: "")) {
+                    refreshApiKey(manga, chapter)
+                    fetch()
+                } else {
+                    Observable.error(error)
+                }
+            }
+    }
+    private val errorRegex = Regex("""(api)?\s*rate\s*limit\s*(excessed)?|api\s*key\s*(invalid)?""")
```

Comment:
As used these parts are useless. Should be equivalent to:
```suggestion
    private val errorRegex = Regex("""\s*rate\s*limit\s*|api\s*key\s*""")
```